### PR TITLE
Handle misdirected requests to OAuth server

### PR DESCRIPTION
**Pull request by enj**
Force kube 1.14 golang.org/x/net

This allows the http2 code to respect "Connection" == "close" to mean sending a GOAWAY and tearing down the TCP connection.

Signed-off-by: Monis Khan <mkhan@redhat.com>

---

Handle misdirected requests to OAuth server

The OAuth server is now resilient against misdirected requests:

1. Limit the number of concurrent streams that each client may have open at a time to the OAuth server to 1.
2. Detect misdirected requests and respond with connection close, 421 status code and HTTP meta refresh.
3. Aggressively close connections that could lead to misdirected requests (basically anything that does not redirect back to the OAuth server).

Signed-off-by: Monis Khan <mkhan@redhat.com>

---

Always set standard security headers for OAuth server

We should always set these headers on all pages of the OAuth server.

Per https://httpwg.org/specs/rfc7540.html#MisdirectedRequest:

    A 421 response is cacheable by default, i.e., unless otherwise
    indicated by the method definition or explicit cache controls
    (see Section 4.2.2 of [RFC7234]).

Thus for 421 we need to set Cache-Control: no-cache (which is part of the standard headers).

Signed-off-by: Monis Khan <mkhan@redhat.com>

---

@openshift/sig-auth @knobunc @ironcladlou @Miciah @deads2k @smarterclayton @jupierce

### CODE REVIEW

**Comment by deads2k** [comment]
I prefer http1.  It's https://github.com/openshift/origin/pull/22515 with a wider `if` plus an env var right?  @smarterclayton your preference?

File: pkg/oauthserver/http2/misdirected_request.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+	// ignore non-TLS requests
+	tls := r.TLS
+	if tls == nil {
+		return false
+	}
+
+	// check if we know the server name requested by the client
+	serverName := tls.ServerName
+	if len(serverName) == 0 {
```
Comment (mrogers950) [inline pkg/oauthserver/http2/misdirected_request.go]: In what situation would this be 0? A non-SNI request?

Comment (enj) [reply pkg/oauthserver/http2/misdirected_request.go]: `ServerName                  string                // server name requested by client, if any (server side only)`

I think it is optional in general?  For example, I believe when the operator hits `/healthz` this value is empty.

**Comment by enj** [comment]
For reference, @jupierce, @Dan and @spadgett all tried to reproduce the issue with a cluster that had this fix and valid certs from Let's Encrypt.  None of them encountered any issue (while being able to readily see the issue on a cluster running master).

**Comment by enj** [comment]
> I prefer http1. It's #22515 with a wider `if` plus an env var right? @smarterclayton your preference?

If the change here is an issue, I updated #22519 to the "minimal" change required: 95efa7d6949bc6ae235de38cda81c81c042d1abe

**Comment by enj** [comment]
#22529

**Pull request closed without merging**